packages/compartment-mapper/src/map-parser.js

```diff
@@ -0,0 +1,294 @@
+   * @param {*} options
+   * @returns {Generator<ReturnType<ModuleTransform>|ReturnType<SyncModuleTransform>, ParseResult, Awaited<ReturnType<ModuleTransform>|ReturnType<SyncModuleTransform>>>}
+   */
+  function* getParserGenerator(
+    bytes,
+    specifier,
+    location,
+    packageLocation,
+    options,
+  ) {
+    /** @type {string} */
+    let language;
+    const extension = parseExtension(location);
+
+    if (
+      !extensionImpliesLanguage(extension) &&
+      has(languageForModuleSpecifier, specifier)
+    ) {
+      language = languageForModuleSpecifier[specifier];
+    } else {
+      language = languageForExtension[extension] || extension;
+    }
+
+    /** @type {string|undefined} */
+    let sourceMap;
+
+    if (has(transforms, language)) {
+      try {
+        ({
+          bytes,
+          parser: language,
+          sourceMap,
+        } = yield transforms[language](
+          bytes,
+          specifier,
+          location,
+          packageLocation,
+          {
+            // At time of writing, sourceMap is always undefined, but keeping
+            // it here is more resilient if the surrounding if block becomes a
+            // loop for multi-step transforms.
+            sourceMap,
+          },
+        ));
+      } catch (err) {
+        throw Error(
+          `Error transforming ${q(language)} source in ${q(location)}: ${err.message}`,
+          { cause: err },
+        );
+      }
+    }
+    if (!has(parserForLanguage, language)) {
+      throw Error(
+        `Cannot parse module ${specifier} at ${location}, no parser configured for the language ${language}`,
+      );
+    }
+    const { parse } = parserForLanguage[language];
+    return parse(bytes, specifier, location, packageLocation, {
+      sourceMap,
+      ...options,
+    });
+  }
+
+  /**
+   * @param {Uint8Array} bytes
+   * @param {string} specifier
+   * @param {string} location
+   * @param {string} packageLocation
+   * @param {*} options
+   * @returns {ParseResult}
+   */
+  const syncParser = (bytes, specifier, location, packageLocation, options) => {
+    return syncTrampoline(
+      getParserGenerator,
+      bytes,
+      specifier,
+      location,
+      packageLocation,
+      options,
+    );
+  };
+
+  /**
+   * @param {Uint8Array} bytes
+   * @param {string} specifier
+   * @param {string} location
+   * @param {string} packageLocation
+   * @param {*} options
+   * @returns {Promise<ParseResult>}
+   */
+  const asyncParser = async (
+    bytes,
+    specifier,
+    location,
+    packageLocation,
+    options,
+  ) => {
+    return asyncTrampoline(
+      getParserGenerator,
+      bytes,
+      specifier,
+      location,
+      packageLocation,
+      options,
+    );
+  };
+
+  // if we have nothing in the moduleTransforms object, then we can use the syncParser.
+  if (keys(moduleTransforms).length === 0) {
+    transforms = syncModuleTransforms;
+    return syncParser;
+  }
+
+  transforms = moduleTransforms;
```

Comment:
AFAIR previeous iteration of this code was using both `[...moduleTransforms, ...syncModuleTransforms]` for the async case. Is this change intentional?

Reply:
it happens in `link`, here: https://github.com/endojs/endo/blob/476069c6df51edf2cfd089ed03e81403bf94751f/packages/compartment-mapper/src/link.js#L301-L310